chapter_recurrent-neural-networks/rnn-scratch.md

```diff
@@ -360,6 +489,34 @@ def train_epoch_ch8(model, train_iter, loss, updater, device,  #@save
     return math.exp(metric[0] / metric[1]), metric[1] / timer.stop()
 ```
 
+```{.python .input  n=170}
+#@tab tensorflow
+def train_epoch_ch8(model, train_iter, loss, updater,  #@save
+                    use_random_iter, vocab, num_hiddens):
+    state, timer = None, d2l.Timer()
+    metric = d2l.Accumulator(2) 
+    params = get_params(len(vocab), num_hiddens)
+    for X, Y in train_iter:
+        if state is None or use_random_iter:
+            state = model.begin_state(batch_size=X.shape[0])
+        else:
+            state = model.begin_state(batch_size=X.shape[0])
+        with tf.GradientTape(persistent=True) as g:
+            g.watch(params)
+            py, state= model(X.numpy(), state, params)
+            y = d2l.reshape(Y.numpy().T, (-1))
+            l = tf.math.reduce_mean(loss(y, py)) 
+        grads = g.gradient(l, params)
+        grads = grad_clipping(grads, 1)
+        updater.apply_gradients(zip(grads, params))
+        
+        # Keras loss by default returns the average loss in a batch
```

Comment:
@terrytangyuan  do I need l_sum or l here?
i.e metric.add(l* d2l.size(y), d2l.size(y)) or metric.add(l_sum* d2l.size(y), d2l.size(y))

Reply:
what about this?

Reply:
@terrytangyuan can you take a look at this?